File: Medium/2026-03-18-3070-Count-Submatrices-with-Top-Left-Element-and-Sum-Less-Than-k/solution.js
class Solution {
    countSubmat(grid, k) {
        const m = grid.length;
        const n = grid[0].length;
        let count = 0;

        for (let right = 0; right < n; right++) {
            const sums = new Array(m).fill(0);
            for (let left = right; left >= 0; left--) {
                for (let row = 0; row < m; row++) {
                    sums[row] += grid[row][left];
                }

                count += this.countLessThanK(sums, k);
            }
        }

        return count;
    }

    countLessThanK(sums, k) {
        let count = 0;
        const prefixSum = [0];
        let currentSum = 0;

        for (const sum of sums) {
            currentSum += sum;
            const target = currentSum - k;

            let left = 0;
            let right = prefixSum.length;

            while (left < right) {
                const mid = Math.floor((left + right) / 2);
                if (prefixSum[mid] <= target) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }

            count += left;
            prefixSum.push(currentSum);
        }

        return count;
    }
}

File: Medium/2026-03-18-3070-Count-Submatrices-with-Top-Left-Element-and-Sum-Less-Than-k/solution.py
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        count = 0
        
        for r1 in range(m):
            row_sum = [0] * n
            for r2 in range(r1, m):
                for c in range(n):
                    row_sum[c] += grid[r2][c]
                
                # Use a hashmap to count the number of subarrays with sum < k
                prefix_sum = {0: 1}
                current_sum = 0
                
                for sum_value in row_sum:
                    current_sum += sum_value
                    # We need to find the number of prefix sums that are less than current_sum - k
                    target = current_sum - k
                    for ps in list(prefix_sum.keys()):
                        if ps <= target:
                            count += prefix_sum[ps]
                    
                    # Update the prefix_sum map
                    if current_sum in prefix_sum:
                        prefix_sum[current_sum] += 1
                    else:
                        prefix_sum[current_sum] = 1
        
        return count